src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,929 @@
+
+[^WangRealSymplecticGroup]:
+    > Wang, Jing and Sun, Huafei and Fiori, Simone:
+    > A Riemannian-steepest-descent approach for optimization on the real symplectic group,
+    > Mathematical Methods in the Applied Sciences, 41(11) pp. 4273-4286, 2018
+    > doi [10.1002/mma.4890](https://doi.org/10.1002/mma.4890)
+"""
+exp(::Symplectic, ::Any...)
+
+function exp!(M::Symplectic, q, p, X)
+    p_star_X = symplectic_inverse_times(M, p, X)
+    q .= p * exp(Array(p_star_X')) * exp(p_star_X - p_star_X')
+    return q
+end
+
+@doc raw"""
+    gradient(M::Symplectic, f, p, backend::RiemannianProjectionBackend;
+             extended_metric=true)
+    gradient!(M::Symplectic, f, p, backend::RiemannianProjectionBackend;
+             extended_metric=true)
+
+Compute the manifold gradient ``\text{grad}f(p)`` of a scalar function
+``f \colon \operatorname{Sp}(2n) \rightarrow ℝ`` at
+``p \in \operatorname{Sp}(2n)``.
+
+The element ``\text{grad}f(p)`` is found as the Riesz representer of the differential
+``\text{D}f(p) \colon T_p\operatorname{Sp}(2n) \rightarrow ℝ`` w.r.t.
+the Riemannian metric inner product at ``p`` [^FioriSimone2011].
+That is, ``\text{grad}f(p) \in T_p\operatorname{Sp}(2n)`` solves the relation
+````math
+    g_p(\text{grad}f(p), X) = \text{D}f(p) \quad\forall\; X \in T_p\operatorname{Sp}(2n).
+````
+
+The default behaviour is to first change the representation of the Euclidean gradient from
+the Euclidean metric to the [`RealSymplecticMetric`](@ref) at ``p``, and then we projecting
+the result onto the correct tangent tangent space ``T_p\operatorname{Sp}(2n, ℝ)``
+w.r.t the Riemannian metric ``g_p`` extended to the entire embedding space.
+
+# Arguments:
+- `extended_metric = true`: If `true`, compute the gradient ``\text{grad}f(p)`` by
+    first changing the representer of the Euclidean gradient of a smooth extension
+    of ``f``, ``∇f(p)``, w.r.t. the [`RealSymplecticMetric`](@ref) at ``p``
+    extended to the entire embedding space, before projecting onto the correct
+    tangent vector space w.r.t. the same extended metric ``g_p``.
+    If `false`, compute the gradient by first projecting ``∇f(p)`` onto the
+    tangent vector space, before changing the representer in the tangent
+    vector space to comply with the [`RealSymplecticMetric`](@ref).
+
+
+[^FioriSimone2011]:
+    > Simone Fiori:
+    > Solving minimal-distance problems over the manifold of real-symplectic matrices,
+    > SIAM Journal on Matrix Analysis and Applications 32(3), pp. 938-968, 2011.
+    > doi [10.1137/100817115](https://doi.org/10.1137/100817115).
+"""
+function gradient(
+    M::Symplectic,
+    f,
+    p,
+    backend::RiemannianProjectionBackend;
+    extended_metric=true,
+)
+    Y = allocate_result(M, gradient, p)
+    return gradient!(M, f, Y, p, backend, extended_metric=extended_metric)
+end
+
+function gradient!(
+    M::Symplectic,
+    f,
+    X,
+    p,
+    backend::RiemannianProjectionBackend;
+    extended_metric=true,
```

Comment:
I am not yet 100% sure this is the way to go, but this is one way to go.
@mateuszbaran what do you think? The difference here is

With extended `extended_metric = false`the embedding (which is not a parameter here) is just the Euclidean space – we first project and then convert from Euclidean metric to Symplectic (This is actually new, but not fast, still the default gradient conversion way)

With `extended_metric=true` the metric conversion comes first (that is Euclidean from default metric to an Extended Symplectic – which is not a metric, but locally projected is fine) and then the projection (that is w.r.t. symplectic metric).

Since this difference only appears very rarely and for now really only here, I think a kwarg would be fine as it is now implemented. What do you think?

Reply:
Yes, it's usually fine to add new keyword arguments to methods for specific manifolds. We can leave it as is, and change it later when we have a better idea. `gradient` interface isn't particularly stable anyway.